rclpy/rclpy/qos_event.py

```diff
@@ -0,0 +1,172 @@
+"""
+QoSRequestedDeadlineMissedInfo = namedtuple(
+    'QoSRequestedDeadlineMissedInfo',
+    ('total_count', 'total_count_change'))
+
+"""
+Payload type for Subscription Liveliness callback.
+
+Mirrors rmw_liveliness_changed_status_t from rmw/types.h
+"""
+QoSLivelinessChangedInfo = namedtuple(
+    'QoSLivelinessChangedInfo',
+    ('alive_count', 'not_alive_count', 'alive_count_change', 'not_alive_count_change'))
+
+"""
+Payload type for Publisher Deadline callback.
+
+Mirrors rmw_offered_deadline_missed_status_t from rmw/types.h
+"""
+QoSOfferedDeadlineMissedInfo = namedtuple(
+    'QoSOfferedDeadlineMissedInfo',
+    ('total_count', 'total_count_change'))
+
+"""
+Payload type for Publisher Liveliness callback.
+
+Mirrors rmw_liveliness_lost_status_t from rmw/types.h
+"""
+QoSLivelinessLostInfo = namedtuple(
+    'QoSLivelinessLostInfo',
+    ('total_count', 'total_count_change'))
+
+
+class SubscriptionEventCallbacks:
+    """Container to provide middleware event callbacks for a Subscription."""
+
+    def __init__(
+        self,
+        *,
+        deadline: Callable[[QoSRequestedDeadlineMissedInfo], None] = None,
+        liveliness: Callable[[QoSLivelinessChangedInfo], None] = None,
+    ) -> None:
+        """
+        Constructor.
+
+        :param deadline: A user-defined callback that is called when a topic misses our
+            requested Deadline.
+        :param liveliness: A user-defined callback that is called when the Liveliness of
+            a Publisher on subscribed topic changes.
+        """
+        self.deadline = deadline
+        self.liveliness = liveliness
+
+
+class PublisherEventCallbacks:
+    """Container to provide middleware event callbacks for a Publisher."""
+
+    def __init__(
+        self,
+        *,
+        deadline: Callable[[QoSOfferedDeadlineMissedInfo], None] = None,
+        liveliness: Callable[[QoSLivelinessLostInfo], None] = None
+    ) -> None:
+        """
+        Constructor.
+
+        :param deadline: A user-defined callback that is called when the Publisher misses
+            its offered Deadline.
+        :param liveliness: A user-defined callback that is called when this Publisher
+            fails to signal its Liveliness and is reported as not-alive.
+        """
+        self.deadline = deadline
+        self.liveliness = liveliness
+
+
+class QoSEventHandler(Waitable):
+    """Waitable type to handle QoS events."""
+
+    def __init__(
+        self,
+        *,
+        callback_group,
+        callback,
+        event_type,
+        parent_handle,
+    ):
+        # Waitable init adds self to callback_group
+        super().__init__(callback_group)
+        self.event_type = event_type
+        self.callback = callback
+
+        self._parent_handle = parent_handle
+        self._event_handle = _rclpy.rclpy_create_event(event_type, parent_handle)
+        self._event_is_ready = False
+        self._event_index = None
+
+    # Start Waitable API
```

Comment:
From here and below, typing is missing.

Reply:
I'm directly implementing the Waitable interface - https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/waitable.py#L68 - which doesn't have typing - I think that adding that would be a separate change from this one.